enterprise/backend/src/metabase_enterprise/semantic_search/index.clj

```diff
@@ -245,12 +246,20 @@
 
 (defn upsert-index!
   "Inserts or updates documents in the index table. If a document with the same
-  model + model_id already exists, it will be replaced."
+  model + model_id already exists, it will be replaced. Parallelizes batch insertion
+  using a thread pool with a configurable thread count (default: 2)."
   [connectable index documents-reducible]
-  (transduce (comp (partition-all *batch-size*)
-                   (map #(upsert-index-batch! connectable index %)))
-             (partial merge-with +)
-             documents-reducible))
+  (cp/with-shutdown! [pool (cp/threadpool (semantic-settings/index-update-thread-count))]
+    (transduce
+     (comp
+      (partition-all *batch-size*)
+      (map #(cp/future pool (upsert-index-batch! connectable index %))))
+     (completing
+      (fn [acc f]
+        (let [result @f]
+          (merge-with + acc result))))
```

Comment:
this is not going to parallelise anything, dereffing in the reducing function will just pull one entry at a time, materalising one future and immediately waiting for it, see 
```
(cp/with-shutdown! [pool (cp/threadpool 4)]
  (transduce
   (map #(cp/future pool (println %) (Thread/sleep 1000) %))
   (completing
    (fn [acc f]
      (let [result @f]
        (conj acc result))))
   []
   [1 2 3 4 5]))
```

Reply:
Yeah, I realized that while testing, still not sure the best way to parallelize while preserving the laziness though

Reply:
I think it could work if we use something like 
```clojure
(defn windowed-reducer [window-size rf]
  (let [window (atom clojure.lang.PersistentQueue/EMPTY)]
    (completing
     (fn [acc f]
       (let [current-window (swap! window conj f)]
         (if (>= (count current-window) window-size)
           (let [oldest (peek current-window)
                 remaining (pop current-window)
                 result @oldest]
             (reset! window remaining)
             (rf acc result))
           acc)))
     (fn [acc] (reduce (fn [acc f] (rf acc @f)) acc @window)))))
```


and 
```
 (windowed-future-reducer (semantic-settings/index-update-thread-count) (partial merge-with +))
```

as reducing function